File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
from collections import deque
import sys
# input = sys.stdin.readline
n,m = map(int,input().split())

number = [list(map(int,input())) for _ in range(n)]
visited = [[False]*m for _ in range(n)]
dx = [1,0,-1,0]
dy = [0,1,0,-1]

def bfs(x,y):
    q = deque()
    q.append((x,y))
    visited[x][y] = True
    while q:
        x1,y1 = q.popleft()
        for i in range(4):
            nx = x1 + dx[i]
            ny = y1 + dy[i]
            if 0<=nx<n and 0<=ny<m:
                if(visited[nx][ny] == False and number[nx][ny] == 1):
                    visited[nx][ny] = True
                    number[nx][ny] = number[x1][y1] + 1 
                    q.append((nx,ny))
    
    return number[n-1][m-1]

    


print(bfs(0,0))

File: 백준/Silver/2178. 미로 탐색/미로 탐색.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

//TIP To <b>Run</b> code, press <shortcut actionId="Run"/> or
// click the <icon src="AllIcons.Actions.Execute"/> icon in the gutter.
public class Main {
    static int n,m;
    static int[][] map;
    static int[] dx={1,0,-1,0};
    static int[] dy={0,1,0,-1};
    static boolean[][] visited;
    static int distance;
    static StringTokenizer st;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        map = new int[n][m];
        visited = new boolean[n][m];
        for(int i = 0; i<n; i++){
            String word = br.readLine();
            for(int j = 0; j<m; j++){
                map[i][j] = word.charAt(j)-'0';
            }
        }
        //BFS
        BFS(0,0);
        System.out.println(map[n-1][m-1]);

    }
    static void BFS(int z,int c){
        visited[z][c] = true;
        Deque<int[]> d = new ArrayDeque<>();
        d.add(new int[]{z,c});
        while(!d.isEmpty()){
            int[] cur = d.poll();
            int x = cur[0];
            int y = cur[1];
            for(int i = 0; i<4; i++){
                int nx = x + dx[i];
                int ny = y + dy[i];
                if(nx<0 || nx>=n || ny<0 || ny>=m) continue;
                if(map[nx][ny] == 1 && !visited[nx][ny]){
                    map[nx][ny] = map[x][y] + 1;
                    visited[nx][ny] = true;
                    d.add(new int[]{nx,ny});
                }
            }
        }
    }



}

